Update deps, modernize, and change everything.

davidlehn opened the pull request:
Going for a simpler API with this and leveraging other tools.  Local tests work, the qr code thing is neat, use on AWS not yet tested, but nodemailer supports it.  Picked using EJS by default since it looks good enough. PUG is the default, but is strange to me. We can change to just about any template system we want.


- **BREAKING**: Switch to `nodemailer` and `email-templates`.
- **BREAKING**: Drop direct SWIG template support.
- **BREAKING**: Drop triggers.
- **BREAKING**: Change API.
  - Changed to async/await.
  - See docs.
- Default to 'ejs' templates.
- Add test app.

davidlehn committed:
Update deps, modernize, and change everything.

- **BREAKING**: Switch to `nodemailer` and `email-templates`.
- **BREAKING**: Drop direct SWIG template support.
- **BREAKING**: Drop triggers.
- **BREAKING**: Change API.
  - Changed to async/await.
  - See docs.
- Default to 'ejs' templates.
- Add test app.

gannan08 commented on README.md:
```diff
@@ -16,124 +16,111 @@ npm install bedrock-mail
-  name: 'My Service'
-  email: 'support@example.com'
+
+// setup default template locals for all mails
```
What does "locals" mean?

davidlehn replied:
It's the template data container terminology used by https://email-templates.js.org/.  We had "mail.vars" before, which aligned with "view.vars", which has since been dropped.  I always thought "vars" was also just as confusing.  What would you suggest instead?

aljones15 replied:
I think `locals` is fine as the `locals` are merged with a template which becomes a full fleshed out HTML email remotely. We might want to add a note or a link to their docs:

`@see https://email-templates.js.org/#/`

or something as when I think of local I think of the local file system first. So a disambiguation might work.

davidlehn replied:
The docs already link to and say it's using email-templates.  Their docs describe how the params like message and locals work.  I had hoped their docs in addition to the simple example would be enough.  I think someone with new eyes on the docs needs to suggest a PR with improved text.  I'd rather not duplicate their docs if possible.

gannan08 commented on lib/index.js:
```diff
@@ -0,0 +1,127 @@
+const nodemailer = require('nodemailer');
+const path = require('path');
+
+const BedrockError = bedrock.util.BedrockError;
```
```suggestion
const {BedrockError} = bedrock.util;
```

gannan08 commented on lib/index.js:
```diff
@@ -0,0 +1,127 @@
+// email client
+let client;
+
+const logger = bedrock.loggers.get('app').child('bedrock-mail');
```
It is my understanding that our convention is to have the logger separated into its own file and pulled in. If we want to break that pattern here, then that's fine.

davidlehn replied:
That should certainly be done to share the logger when multiple files are logging.  But in this case, it's just one main file that does logging.  I figured we didn't need to bother for the simple one file use case.

aljones15 replied:
I made an issue: https://github.com/digitalbazaar/bedrock-mail/issues/5

davidlehn committed:
Use object destructuring.

Co-Authored-By: Ganesh Annan <gannan08@users.noreply.github.com>

davidlehn committed:
Improve comments.

davidlehn committed:
Add use() and verify() APIs.

- Allow transport to be configured and checked at runtime.

davidlehn committed:
Improve docs.

davidlehn committed:
Add more logging options.

- Add more fine grained options for development use logging.
- Can output headers, text, and/or html.
- Always output basic debugging info.

davidlehn committed:
Add main --mail-{preview,send,log} options.

- Useful to easily override these during development.

davidlehn committed:
Add --mail-to option to override mail target.

- Useful during development to send emails from test account emails to
  another real account.

davidlehn committed:
Update dependencies.

davidlehn committed:
Use object spread.

davidlehn committed:
Add FIXME.

davidlehn committed:
Log response.

davidlehn committed:
Update changelog.

davidlehn committed:
Remove redundant --to option.  Use --mail-to.

davidlehn committed:
Add template test option.

davidlehn committed:
Add "include-test" template.

davidlehn committed:
Update docs.

davidlehn committed:
Add reference.

davidlehn committed:
Fix style.

davidlehn committed:
Add CLI option docs.

aljones15 commented on test/index.js:
```diff
@@ -0,0 +1,115 @@
+  console.log('emitting event');
+  // use special event for main test, else use generic event
+  const eventName =
+    bedrock.program.tempalte === 'mail-test.notify' ?
```
this is none of my business... I mean you can do what you want but shouldn't this be `template` or is this some new term used in the HTML mail business I'm not familiar with?

aljones15 commented on lib/config.js:
```diff
@@ -1,23 +1,76 @@
-config.mail.events = [];
+
+// transport
+// can setup in config or at runtime with use('transport', ...);
```
good docs here! THANK YOU!

aljones15 reviewed (changes requested):
Good work. You're awesome. PR DENIED.

davidlehn committed:
Fix typo.

aljones15 reviewed (approved):
this looks good to me. cookies.

davidlehn committed:
Async fixes.

- Note in docs APIs are async.
- Add await to `use` call.

davidlehn committed:
Allow verify on any transports where available.

davidlehn committed:
Add `--transport-check` feature.

- Allow program to quit early. Useful for debugging.

davidlehn committed:
Improve error handling.

davidlehn committed:
Update changelog.

davidlehn committed:
Add an await.

davidlehn committed:
Add views root logging.

mattcollier commented:
@davidlehn what's the status on this PR?

davidlehn commented:
@mattcollier I was waiting for a bit more testing. Got that and seems to be working enough for a release now. I'll do so.
